### ISSUE TITLE
As a data editor, I want to view and edit library items as works and associated editions so that I can have events related to the same item grouped but still document known editions.

### ISSUE THREAD

**Issue by rlskoeser** [Princeton-CDH/mep-django#350]
## notes for testing
- Bibliography *Item* is now renamed *Work* in admin, should be listed on main dashboard
- When you edit Work you have an optional inline for associated Editions 
- Works with matching work URIs, matching authors, and similar titles, should be merged
  - all creators should be copied over to the merged work
  - all subjects and genres should be combined on the merged work
  - attributes should be copied over on the merged work if not already present (mep id, year, ebook url, format)
  - notes from all merged works should be combined
- Works with matching work URIs where titles do not match authors or titles are too dissimilar should *not* be merged
- merge should be noted in the notes - date & list of mep ids or database ids for items without an mep id
- if titles vary, the notes should include TITLEVAR
- if years vary, the merged work should use the earliest year 
- there should be an admin log entry on the merged works documenting when the merge was done
- please also check examples for all of the following cases (list from brainstorming session 2019/05/29)
   - merged incorrectly (I think this means previous bad merge?)
   - merged correctly, 2 different known editions
   - periodicals
   - multivolume works
   - works in multiple languages
   - works with multiple types of creators

## dev notes
* data migrations to do:
  * [x] rename `Item` to `Work` everywhere
  * [x] split `Work` fields and create `Edition` model
  * [x] modify generic Event to link to both Work and Edition
  * [x] update Work admin so associated Editions can be edited as an inline 
  * [x] update Event, Borrow, and Purchase admin edit to allow associating Edition
  * [x] group item by OCLC work URI into Works
  * ~create editions associated with work based on edition URIs~
  * [x] migrate all events associated with items to Work ~and optional edition if edition is known~
  * ~check if borrow event notes field is actionable for edition information (periodicals?)~

**Comment by clmahoney** [comment]
@jkotin can you look through this as well? Mainly this point: 
- please also check examples for all of the following cases (list from brainstorming session 2019/05/29)
--> merged incorrectly (I think this means previous bad merge?)
--> merged correctly, 2 different known editions
--> periodicals
--> multivolume works
-->works in multiple languages
-->works with multiple types of creators

**Comment by jkotin** [comment]
Thanks Cate.  I haven't been testing, so please send me step-by-step instructions about what to look at and for, and where.  J

**Comment by clmahoney** [comment]
@jkotin OK! RSK wrote very detailed instructions at the top of the page. I can test everything, but it would be helpful for me to know if you have specific problematic items in mind (known multi-language and also two different known editions). If you're busy, I can do it, but if you have any off the top of your head! Thank you!

**Comment by jkotin** [comment]
Hi Cate -- I need very basic pointers: what page?  What am I testing?  I am happy to do it.  So far I've been leaving this to you, so I'm totally out of the loop. J

**Comment by rlskoeser** [comment]
@clmahoney thanks for documenting the list of things you checked - for things that require looking at specific items, please note how you tested - e.g. examples of *specific* works you tested for title variance or things not being merged. 

The details of how you're testing can be really helpful - it might make it obvious if you and I have different expectations or understandings of something, and it's also valuable to refer back to if we run into a problem later on that we didn't catch in testing (those details might give us enough information to understand what's going on and why we missed it, which will make it easier to fix).

**Comment by jkotin** [comment]
I started to look at the test site today.  It looks good.  The are a number of items that should be merged, yet are not:

A Gallery of Women (6912, 239)
A Great Sea Mystery (5958, 2820)
A Modern Hero (4983, 3978)
A New American History (8436, 5094)
A Novelette and Other Prose (1715, 1882)
A Portrait of the Artist as a Young Man (4339, 3240)
A Selection from the Letters of Lewis Carroll (3946, 1130)
A Story Anthology, 1931-1933 (3833, 5463)
A Tale of a Tub (1767, 7239, 7219)
Alice's Adventures in Wonderland (8008, 2163)
Experiment in Autobiography (8419, 43)
Fathers and Children (1957, 976)
The Brothers Karamazov (2969, 4295)

(You can see how I checked.) But I think there are explanations for why these merges didn't happen. Some items are recorded as not having an OCLC record, and some have minor variations in their titles.)  Will there be a way to manually merge such items?  Or will we manually enter work uri and automate the merge process?

I also made a list of items that seemed to merge correctly. But I would be grateful if Cate to verify by looking at the lending library cards:

A Life of William Shakespeare by Lee (8270)
A Life of William Shakespeare by Quincy (8253)
Alexander Pope by Stephen (8329)
Alexander Pope by Sitwell (4768)

Are there other issues I should be looking for?  Should I check to see if the work uris are correct?

Additional notes:

The French translation of Ulysses was merged into 6351.  We may want that to happen in general (i.e. have all the translations of a single work share a work uri), but not for Ulysses.

**Comment by clmahoney** [comment]
Thank you Josh!! 
I've checked the ones you wanted me to: 8270 is correctly merged (both Lee), 8253 is correctly merged (JQ Adams), 8329 is correctly merged (Pope by Stephen). I suspect that all of the merges on 4768 (Pope by Sitwell) are correct, but 3 of them mention simply "Alexander Pope" as their title although the events are all from 1930, when the Sitwell was published.

**Comment by clmahoney** [comment]
I can manually merge books that were not merged; in some of my work so far I've been doing that already.

**Comment by jkotin** [comment]
Excellent.  That all sounds perfect.

J

**Comment by jkotin** [comment]
Excellent.  Let me know if any more testing (or exploration) would be useful.  

Re: Portrait issue: Is there an easy way to see if similar titles have different work uris?

I'm not worried about the French Ulysses -- it's an exception.  We can always demerge later, correct?  Or delete a borrowing event and enter a new one?

J

**Comment by clmahoney** [comment]
I realize I should show my testing for specifics more: 
- For TITLEVAR, an example that looks good is: 5448, which is The Old Possum's Book of Practical Cats. Variant titles: Book of Practical Cats (on Bassiano's card). 
- For PERIODICALs, an example is 6299, which is New Statesman + Nation (the periodicals merged after 1931): all the borrows are after 1931, which is good!

**Comment by clmahoney** [comment]
- For MULTIVOLUME, 8447 is Lord Byron's Correspondence, and I see that volumes 1 and 2 are merged here. We note the different volumes in the notes section.

**Comment by clmahoney** [comment]
I'm unclear as to what "merged incorrectly" means, however, or how to check that.

**Comment by rlskoeser** [comment]
@clmahoney thank you for documenting specific examples for what you tested, super helpful.

 don't know what we meant by "merged incorrectly" either! It sounds like you've looked at a lot of records in your testing, have you come across anything that was merged that shouldn't have been?

@jkotin I think you're right, it should be pretty easy to handle the French translation of Ulysses manually later on.

To your question about identifying other similar titles with work URIs: I figured out I could check this with the CSV download and OpenRefine, clustering records by title and checking the URIs within a cluster. Here's what I found:
- 5 different Items for Hamlet, two have work URIs and they are different
- Collected Poems of H.D.: 4 records, all with the same work URI; not sure why this one didn't get merged
- Studies in the Psychology of Sex: 4 records, two different work URIs 
- The New Poetry: An Anthology of Twentieth-Century Verse in English: 4 records, all with the same work URI; not sure why this one didn't get merged either

I'm not sure how to check easily for similar titles with a large number of events; the ones I'm finding don't have a lot of events, so if we have to do manual clean up it doesn't seem like it will be terrible.

Looking at in OpenRefine gave me the idea to cluster by work URI, which also seems to be a useful test. I see 7 items by Chekhov that have different titles and shouldn't have been merged, so that's good. I see a pair of _Poems_, one by Poe and one by Rossetti that have the same work URI - not merged, so that seems good. Very quickly I'm getting to where there are only 2 items with the same URI and one or two events, so it doesn't seem like there are a lot of problems here. @clmahoney you might want to review this list (cluster by work URI) to see if there's anything you want to clean up before we do the merge in production.

It does look like I need to look at my code again to check why some of these _didn't_ merge when it looks like they should have.

**Comment by jkotin** [comment]
Thank you Rebecca.  This all makes sense to me.

**Comment by clmahoney** [comment]
I think I can close this now?

**Comment by rlskoeser** [comment]
I finally had time to check the titles that didn't merge that I wanted to follow up, and everything seemed reasonable - the 'ok to merge' said these were not ok to merge:
- the H.D. titles weren't merge because my logic for normalizing initials with and without spaces didn't account for initials being at the end of the title; I've made a small adjustment to handle that
- the poetry anthologies didn't merge because one included an *an* and one did not; I added that to my minimal list of stop words to remove when checking titles.

@clmahoney let me know if you want me to fix the Portrait URIs or you want me to, and if you want to look at any of the other possible clean up before we run the merge (we can discuss at our meeting today). Other than that, if you're comfortable with cleaning up anything that isn't merged automatically after we run this migration, then I think we're good!
